### PR TITLE
Only pass the use_kv_cache True to generator

### PR DESCRIPTION
**Pull request by yafshar**
In case of not using use_kv_cache, the False value causes performance issue

# What does this PR do?

This PR addresses a performance issue caused by setting the `use_cache` argument to `False` by default for all model classes when not passing `--use_kv_cache`. 

The `use_cache` parameter should only be set to `True` when explicitly requested, as its default setting to `False` leads to performance degradation. This PR restores the original behavior, making PR #1343 unnecessary. 

Without the fix

```sh
>>> python3 run_pipeline.py  --model_name_or_path llava-hf/llava-1.5-7b-hf  --use_hpu_graphs --bf16

09/26/2024 20:56:05 - INFO - __main__ - time = 1933.194888199796ms, Throughput (including tokenization) = 20.691136855450857 tokens/second
```

With this fix

```sh
>>> python3 run_pipeline.py  --model_name_or_path llava-hf/llava-1.5-7b-hf  --use_hpu_graphs --bf16

09/26/2024 20:54:54 - INFO - __main__ - time = 942.6103481997416ms, Throughput (including tokenization) = 103.9666073973904 tokens/second
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: tests/test_image_to_text_example.py [hunk 1]
```diff
@@ -57,6 +57,7 @@ def _test_image_to_text(
         f"--model_name_or_path {model_name}",
         f"--batch_size {batch_size}",
         "--max_new_tokens 20",
+        "--use_kv_cache",
```
Comment (vidyasiv) [inline tests/test_image_to_text_example.py:60]: Previous CI did not enable this, do we see similar perf without this first? I think clip doesnt use kv cache so it likely won't help that case.
Alternatively we can add also flash_attention options and see most optimized perf (if it aligns with README)

**Comment by regisss** [comment]
Running
```
python3 run_pipeline.py  --model_name_or_path llava-hf/llava-1.5-7b-hf  --use_hpu_graphs --bf16
```
I do get a better throughput but the generated text is bad:
```
09/26/2024 22:12:23 - INFO - __main__ - result = [[{'generated_text': 'USER:  \nWhat is shown in this image? ASSISTANT: The\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ
\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\n'}]]      
09/26/2024 22:12:23 - INFO - __main__ - time = 971.51561062783ms, Throughput (including tokenization) = 100.87331477532173 tokens/second
```
Do you get the same @yafshar ?

**Comment by kaixuanliu** [comment]
> Running
> 
> ```
> python3 run_pipeline.py  --model_name_or_path llava-hf/llava-1.5-7b-hf  --use_hpu_graphs --bf16
> ```
> 
> I do get a better throughput but the generated text is bad:
> 
> ```
> 09/26/2024 22:12:23 - INFO - __main__ - result = [[{'generated_text': 'USER:  \nWhat is shown in this image? ASSISTANT: The\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ
> \nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\nЪ\n'}]]      
> 09/26/2024 22:12:23 - INFO - __main__ - time = 971.51561062783ms, Throughput (including tokenization) = 100.87331477532173 tokens/second
> ```
> 
> Do you get the same @yafshar ?

@regisss , I meet same wrong output. This PR will solve it, #1371

**Comment by vidyasiv** [comment]
@regisss  , @jiminha 
- Updated to have script support chat template optionally as perf dips with it (for `llava-hf/llava-1.5-7b-hf `, with chat template ~87 tokens/sec and without ~91 tokens/sec)
- I made an update for pad value 0 based on https://github.com/huggingface/optimum-habana/pull/1371 (credit to @kaixuanliu) , without this change I see zero throughput for test: `tests/test_image_to_text_example.py::test_image_to_text_bf16[token0-llava-hf/llava-1.5-13b-hf-1-16.704731010481538]`, i.e `python3 run_pipeline.py  --model_name_or_path llava-hf/llava-1.5-13b-hf  --use_hpu_graphs --bf16`.
- I saw the same perf/accuracy as you did for `python3 run_pipeline.py  --model_name_or_path llava-hf/llava-1.5-7b-hf  --use_hpu_graphs --bf16`, perf drops(97 ->91) but accuracy better with pad value 0 for model_kwargs[other_inputs]. 
- The pad value was introduced by input embeddings PR, I tested Mistral with and without input embeddings and noticed perf improvement and PR accuracy unaffected.
```
root@800672-7:~/synopsis/yafshar/examples/text-generation# python run_generation.py --model_name_or_path mistralai/Mistral-7B-Instruct-v0.2  --input_embeds --batch_size 6
<snip>
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework for distributed training and inference, developed by Meta. It is designed to be efficient on large-scale systems, and supports both PyTorch and TensorFlow models. DeepSpeed includes several components:\n\n1. **Model Parallelism**: DeepSpeed allows you to split your model across multiple GPUs or TPUs, which can help reduce memory usage and improve training speed.\n2. **Data Parallelism**: DeepSpeed supports standard data parallelism, where the model is re',)

input 2: ('He is working on',)
output 1: ('He is working on a new project, which is a 3D platformer game called "The Lost Temple". The game is set in ancient Egypt and follows the story of a young adventurer named Khepri, who embarks on a journey to find the legendary Lost Temple and uncover its secrets. The game features beautiful 3D graphics, challenging puzzles, and exciting platforming action.\n\nThe development of the game is currently in the early stages, with the team focusing on creating the game',)

input 3: ('He has a',)
output 1: ('He has a point.\n\n> The U.S. government has spent $1.5 trillion on wars in Iraq and Afghanistan, and the cost of caring for veterans of those wars is projected to reach $1 trillion over the next decade.\n>\n> But the U.S. Department of Veterans Affairs has spent just $1.5 million on a program to help veterans get jobs, according to a report by the Government Accountability Office.\n>\n> The Veterans Ret',)

input 4: ('He got all',)
output 1: ('He got all the way to the end of the line, but he couldn\'t get on the train.\n\nThe train was full, and there was a sign on the platform that read, "No standing room available."\n\nBut the man was determined. He had been traveling all day and night to get to this city, and he wasn\'t going to let a little thing like a full train stop him.\n\nSo he did what any resourceful traveler would do. He looked around',)

input 5: ('Everyone is happy and I can',)
output 1: ('Everyone is happy and I can’t help but feel like I’m missing something.\n\nI’ve been in a relationship for a year and a half now, and I’ve never felt this way before. I’ve never felt so content and at peace with someone. We’ve had our ups and downs, but we’ve always managed to work through them. We communicate well, we make each other laugh, and we support each other through thick and thin.\n\nBut sometimes, when',)

input 6: ('The new movie that got Oscar this year',)
output 1: ('The new movie that got Oscar this year is called “The Shape of Water” and it’s a beautiful, magical, and romantic film. It’s a story about a mute woman, Elisa, who works as a cleaning lady at a secret government laboratory in the 1960s. She discovers a mysterious creature in a water tank and falls in love with it.\n\nThe movie is visually stunning, with a lot of attention to detail in the production design, costumes, and cinemat',)


Stats:
-----------------------------------------------------------------------------------
Input embeds
Throughput (including tokenization) = 171.23435324319036 tokens/second
Memory allocated                    = 28.1 GB
Max memory allocated                = 28.56 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 10.896730984095484 seconds
-----------------------------------------------------------------------------------

root@800672-7:~/synopsis/yafshar/examples/text-generation# python run_generation.py --model_name_or_path mistralai/Mistral-7B-Instruct-v0.2  --batch_size 6
<snip>
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework for distributed training and inference, developed by Meta. It is designed to be efficient on large-scale systems, and supports both PyTorch and TensorFlow models. DeepSpeed includes several components:\n\n1. **Model Parallelism**: DeepSpeed allows you to split your model across multiple GPUs or TPUs, which can help reduce memory usage and improve training speed.\n2. **Data Parallelism**: DeepSpeed supports standard data parallelism, where the model is re',)

input 2: ('He is working on',)
output 1: ('He is working on a new project, which is a 3D platformer game called "The Lost Temple". The game is set in ancient Egypt and follows the story of a young adventurer named Khepri, who embarks on a journey to find the legendary Lost Temple and uncover its secrets. The game features beautiful 3D graphics, challenging puzzles, and exciting platforming action.\n\nThe development of the game is currently in the early stages, with the team focusing on creating the game',)

input 3: ('He has a',)
output 1: ('He has a point.\n\n> The U.S. government has spent $1.5 trillion on wars in Iraq and Afghanistan, and the cost of caring for veterans of those wars is projected to reach $1 trillion over the next decade.\n>\n> But the U.S. Department of Veterans Affairs has spent just $1.5 million on a program to help veterans get jobs, according to a report by the Government Accountability Office.\n>\n> The Veterans Ret',)

input 4: ('He got all',)
output 1: ('He got all the way to the end of the line, but he couldn\'t get on the train.\n\nThe train was full, and there was a sign on the platform that read, "No standing room available."\n\nBut the man was determined. He had been traveling all day and night to get to this city, and he wasn\'t going to let a little thing like a full train stop him.\n\nSo he did what any resourceful traveler would do. He looked around',)

input 5: ('Everyone is happy and I can',)
output 1: ('Everyone is happy and I can’t help but feel like I’m missing something.\n\nI’ve been in a relationship for a year and a half now, and I’ve never felt this way before. I’ve never felt so content and at peace with someone. We’ve had our ups and downs, but we’ve always managed to work through them. We communicate well, we make each other laugh, and we support each other through thick and thin.\n\nBut sometimes, when',)

input 6: ('The new movie that got Oscar this year',)
output 1: ('The new movie that got Oscar this year is called “The Shape of Water” and it’s a beautiful, magical, and romantic film. It’s a story about a mute woman, Elisa, who works as a cleaning lady at a secret government laboratory in the 1960s. She discovers a mysterious creature in a water tank and falls in love with it.\n\nThe movie is visually stunning, with a lot of attention to detail in the production design, costumes, and cinemat',)


Stats:
-----------------------------------------------------------------------------------
Input tokens
Throughput (including tokenization) = 55.172114129603564 tokens/second
Memory allocated                    = 28.11 GB
Max memory allocated                = 28.42 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 41.56314965011552 seconds
-----------------------------------------------------------------------------------
```
I looked up the pad_token_id for the models to see if that can explain perf change:
- https://huggingface.co/llava-hf/llava-1.5-13b-hf/blob/main/generation_config.json : `pad_token_id: 32001` (perf change)
- https://huggingface.co/llava-hf/llava-1.5-7b-hf/blob/main/generation_config.json: `pad_token_id: 32001` (perf change)
- https://huggingface.co/llava-hf/llava-v1.6-mistral-7b-hf/blob/main/generation_config.json (no pad_token_id in file but debug print shows 32001) (on the edge of perf change, i.e sometimes fails)
- https://huggingface.co/llava-hf/llava-v1.6-vicuna-7b-hf/blob/main/generation_config.json ` pad_token_ids: 0` (no perf change)

With this PR, all 10 tests pass:
```
pytest -s -v  tests/test_image_to_text_example.py
====== 10 passed in 1133.14s (0:18:53) ======
```
Let me know if you'd like me to test further

File: optimum/habana/transformers/generation/utils.py [hunk 2]
```diff
@@ -957,7 +957,7 @@ def generate(
                             model_kwargs[other_inputs] = torch.nn.functional.pad(
                                 model_kwargs[other_inputs],
                                 (0, generation_config.max_new_tokens),
-                                value=generation_config.pad_token_id,
+                                value=0,
```
Comment (yafshar) [inline optimum/habana/transformers/generation/utils.py:960]: @vidyasiv I am not sure setting the value to `0` is correct. What if a model uses a different padding?

Comment (vidyasiv) [reply optimum/habana/transformers/generation/utils.py:960]: This was introduced in PR: https://github.com/huggingface/optimum-habana/pull/1276. So it used to be 0 while other paddings above it used the pad_token. It is possible this didnt need to be updated. I have noted how it fails accuracy for the models where padding is non zero above.

Comment (yafshar) [reply optimum/habana/transformers/generation/utils.py:960]: I think we should look at that more accurately. My first impression is that there might be inconsistency in the code that caused these issues later.

**Review by tthakkal — APPROVED** [review]

File: optimum/habana/transformers/generation/utils.py (hunk 2 above)
Comment (vidyasiv) [reply optimum/habana/transformers/generation/utils.py:960]: @zongwave please help review this part.

**Comment by vidyasiv** [comment]
@regisss  I can leave out the accuracy fix for now unless the padding change makes sense to you. Let me know what you think.

File: optimum/habana/transformers/generation/utils.py (hunk 2 above)
Comment (jiminha) [reply optimum/habana/transformers/generation/utils.py:960]: Talked with @ssarkar2  Attention_mask value is either 0/1, so it will go wrong if we put non-zero pad token id.  It should be set to 0 here.

Comment (vidyasiv) [reply optimum/habana/transformers/generation/utils.py:960]: PR where 0 was set: https://github.com/huggingface/optimum-habana/pull/488

**Comment by vidyasiv** [comment]
#### Tested models added a part of #1292
```
python3 run_pipeline.py     --model_name_or_path llava-hf/llava-v1.6-34b-hf     --use_hpu_graphs     --bf1
Output:
09/28/2024 01:41:26 - INFO - __main__ - result = [[{'generated_text': '<|im_start|> user\n\n What is shown in this image?\n<|im_start|> assistant\nThe image displays a radar chart, also known as a spider chart, which is a graphical method of displaying multivariate data in the form of a two-dimensional chart of three or more quantitative variables represented on axes starting from the same point.\n\nIn this particular chart, there are several datasets represented by different colors and labels. Each dataset is plotted along multiple axes, which are labeled with various metrics such as "MM-Vet," "LLaVa-Bench," "Seed'}]]
09/28/2024 01:41:26 - INFO - __main__ - time = 96303.48922722042ms, Throughput (including tokenization) = 1.1110708538040819 tokens/second
```
```
python3 run_pipeline.py \
    --model_name_or_path llava-hf/llama3-llava-next-8b-hf \
    --use_hpu_graphs \
    --bf16
Output:
09/28/2024 01:46:07 - INFO - __main__ - result = [[{'generated_text': 'user\n\n\nWhat is shown in this image?assistant\n\n\n\nThe image shows a radar chart, also known as a spider chart or a web chart, which is a type of graph used to display multivariate data in the form of a two-dimensional chart of three or more quantitative variables represented on axes starting from the same point. Each axis represents a different variable, and the values are plotted along each axis and connected to form a polygon.\n\nIn this particular radar chart, there are several axes labeled with different variables, such as "MM-Vet," "'}]]
09/28/2024 01:46:07 - INFO - __main__ - time = 1337.3435568064451ms, Throughput (including tokenization) = 68.04534222851944 tokens/second
```

File: optimum/habana/transformers/generation/utils.py (hunk 2 above)
Comment (regisss) [reply optimum/habana/transformers/generation/utils.py:960]: Yes, both attention masks and token type ids should have values in 0 and 1. So padding with `pad_token_id` was wrong. Going to merge #1371.

**Review by regisss — APPROVED** [review]
LGTM, thanks for fixing that!

Is everybody okay with merging this one?

**Review by vidyasiv — APPROVED** [review]
